stan/math/gpu/multiply.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef STAN_MATH_GPU_MULTIPLY_HPP
+#define STAN_MATH_GPU_MULTIPLY_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/matrix_gpu.hpp>
+#include <stan/math/gpu/kernels/scalar_mul.hpp>
+#include <stan/math/gpu/kernels/matrix_multiply.hpp>
+#include <Eigen/Dense>
+
+namespace stan {
+namespace math {
+/**
+ * Multiplies the specified matrix on the GPU
+ * with the specified scalar.
+ *
+ * @param A matrix
+ * @param scalar scalar
+ * @return matrix multipled with scalar
+ */
+inline matrix_gpu multiply(matrix_gpu& A, double scalar) {
+  matrix_gpu temp(A.rows(), A.cols());
+  if (A.size() == 0)
+    return temp;
+  try {
+    opencl_kernels::scalar_mul(cl::NDRange(A.rows(), A.cols()), temp.buffer(),
+                               A.buffer(), scalar, A.rows(), A.cols());
+  } catch (const cl::Error& e) {
+    check_opencl_error("multiply scalar", e);
+  }
+  return temp;
+}
+
+/**
+ * Multiplies the specified matrix on the GPU
+ * with the specified scalar.
+ *
+ * @param scalar scalar
+ * @param A matrix
+ * @return matrix multipled with scalar
+ */
+inline matrix_gpu multiply(double scalar, matrix_gpu& A) {
+  return multiply(A, scalar);
+}
+
+/**
+ * Computes the product of the specified GPU matrices.
+ *
+ * Computes the matrix multiplication C[M, K] = A[M, N] x B[N, K]
+ *
+ * @param A first matrix
+ * @param B second matrix
+ * @return the product of the first and second matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ *   number of columns in A and rows in B do not match
+ */
+inline matrix_gpu multiply(matrix_gpu& A, matrix_gpu& B) {
+  check_size_match("multiply (GPU)", "A.cols()", A.cols(), "B.rows()",
+                   B.rows());
+  matrix_gpu temp(A.rows(), B.cols());
+  if (temp.size() == 0)
+    return temp;
+  int local = base_opts["WG_SIZE_MULT"];
```

Comment:
I dont like this. We need to be able to check this at compile time.

Reply:
`base_opts` is defined inside a anonymous namespace which makes it local inside a translation unit. But based on your other comment this will get moved to the opencl_context anyways.
```
namespace stan {
namespace math {
namespace {
```

Reply:
Woops, I didn't realize that was global to the entire translation unit - I thought just the file! Eep. While you're in there would you also mind making that one `internal` instead and maybe naming the stuff better? Or I can do that in this or another PR if it's easier.

Reply:
What if I moved it to the `stan::math::opencl_kernels` namespace? Since this will only contain the compile_kernel function (the base_opts will be moved to context).

I would want to avoid 
```
public:
  kernel_functor(const char* name, const char* source)
      : kernel_(stan::math::internal::compile_kernel(name, source)) {}
```

Reply:
I am not sure at which stuff you are aiming at that you say should be named better? The constants in `base_opts or the variables inside this matrix_multiply?

Reply:
the opencl_kernels namespace works for me for this stuff. Re: names - I just meant the `compile` function and `base_opts` etc that were exposed to the entire translation unit (all of Stan Math is one translation unit typically, so that was just totally global, haha). My bad.